crates/service-protocol/src/discovery.rs

```diff
@@ -141,14 +142,34 @@ pub enum ServiceDiscoveryError {
     Descriptor(#[from] DescriptorError),
 
     // Network related retryable errors
-    #[error("retry limit exhausted. Last bad status code: {0}")]
+    #[error("bad status code: {0}")]
     #[code(META0003)]
     BadStatusCode(u16),
-    #[error("retry limit exhausted. Last client error: {0}")]
+    #[error("client error: {0}")]
     #[code(META0003)]
     Client(#[from] ServiceClientError),
 }
 
+impl ServiceDiscoveryError {
+    /// Retryable errors are those which can be caused by transient faults and where
+    /// retrying can succeed.
+    pub fn is_retryable(&self) -> bool {
+        match self {
+            ServiceDiscoveryError::BadStatusCode(status) => matches!(
+                StatusCode::from_u16(*status).expect("should be valid status code"),
+                StatusCode::REQUEST_TIMEOUT
+                    | StatusCode::TOO_MANY_REQUESTS
+                    | StatusCode::INTERNAL_SERVER_ERROR
+                    | StatusCode::BAD_GATEWAY
+                    | StatusCode::SERVICE_UNAVAILABLE
+                    | StatusCode::GATEWAY_TIMEOUT
+            ),
+            ServiceDiscoveryError::Client(client_error) => client_error.is_retryable(),
+            _ => false,
```

Comment:
It'd be helpful for our future selves if you write a comment here about why the default case is `false` vs why it's `true` in SdkError.

Reply:
I will spell the `ServiceDiscoveryError` variants that are not retryable out to make it clear that newly added variants need to be reasoned about.

Reply:
I've pushed a fixup to list the cases explicitly.